src/validation.cpp

```diff
@@ -1130,15 +1130,35 @@ bool ReadRawBlockFromDisk(std::vector<uint8_t>& block, const CBlockIndex* pindex
 
 CAmount GetBlockSubsidy(int nHeight, const Consensus::Params& consensusParams)
 {
-    if (Params().NetworkIDString() != CBaseChainParams::REGTEST)
-        return consensusParams.baseBlockSubsidy;
+    CAmount nSubsidy = consensusParams.baseBlockSubsidy;
+
+    if (Params().NetworkIDString() != CBaseChainParams::REGTEST || consensusParams.EunosHeight == 1)
+    {
+        if (nHeight >= consensusParams.EunosHeight)
+        {
+            nSubsidy = consensusParams.newBaseBlockSubsidy;
+            int reductions = (nHeight - consensusParams.EunosHeight) / consensusParams.emissionReductionPeriod;
+            CAmount reductionAmount;
+            for (; reductions > 0; --reductions)
+            {
+                reductionAmount = (nSubsidy * consensusParams.emissionReductionAmount) / 100000;
+                if (!reductionAmount) {
+                    nSubsidy = 0;
+                    break;
+                }
+
+                nSubsidy -= reductionAmount;
+            }
+        }
```

Comment:
Could we initialize a map with `{hight, subsidy}` (especially on `(nHeight - consensusParams.EunosHeight) % consensusParams.emissionReductionPeriod == 0`) on first call, then bound to height? I mean we don't need to calculated it over and over again.

Reply:
It sounds like a lookup table

Reply:
Good idea. Implemented here as a map against the reduction count. https://github.com/DeFiCh/ain/pull/328/commits/bb0081f9d2bb389bca9e7208c1e0bf4a3cbb648a